### ISSUE TITLE
Seemingly unbounded memory consumption of a new empty replica

### ISSUE THREAD

**Issue by fillest** [ClickHouse/ClickHouse#41065]
The version is 22.7.1.2484 (docker).
We had multiple shards, each ~3TiB on disk. I added a new fresh empty replica for one of the shards. The replication started. I expected moderate or at least controlled memory usage, but it once hit some limit (looks like `max_server_memory_usage` - though it exceeded it for ~10GiB) and spammed "Memory limit (total) exceeded" error. I suspect that this is not healthy, because, for example, I also see warnings about probably-failed fetches and so on. `max_server_memory_usage` can trigger errors in random places, if I understand correctly (which is risky).
The network bandwidth is 1G and it was fully saturated most of the time. The local "disk" is HDD raid10 - but it was resyncing.
So my suspicion is: in this situation Clickhouse periodically downloads parts much faster than it writes to disk, so the memory consumption grows and its growth looks unbounded.
Lowering `background_fetches_pool_size` (I'm not sure it will help though) will probably result in under-utilization and requires time-consuming trial-and-error tuning. So maybe it would be better to add a setting to limit the memory space for downloaded parts (so the downloader will "block" while the limit is hit and will not spam errors), what do you think?

**Error message and/or stacktrace**
```
2022.09.05 21:11:21.598141 [ 46 ] {} <Error> MergeFromLogEntryTask: virtual bool DB::ReplicatedMergeMutateTaskBase::executeStep(): Code: 241. DB::Exception: Memory limit (total) exceeded: would use 116.50 GiB (attempt to allocate chunk of 4578069 bytes), maximum: 107.00 GiB. OvercommitTracker decision: Memory overcommit isn't used. OvercommitTracker isn't set.. (MEMORY_LIMIT_EXCEEDED), Stack trace (when copying this message, always include the lines below):
0. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, bool) @ 0xba37dda in /usr/bin/clickhouse
1. DB::Exception::Exception<char const*, char const*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, long&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string_view<char, std::__1::char_traits<char> > >(int, fmt::v8::basic_format_string<char, fmt::v8::type_identity<char const*>::type, fmt::v8::type_identity<char const*>::type, fmt::v8::type_identity<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >::type, fmt::v8::type_identity<long&>::type, fmt::v8::type_identity<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >::type, fmt::v8::type_identity<std::__1::basic_string_view<char, std::__1::char_traits<char> > >::type>, char const*&&, char const*&&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&&, long&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&&, std::__1::basic_string_view<char, std::__1::char_traits<char> >&&) @ 0xba4fff7 in /usr/bin/clickhouse
2. MemoryTracker::allocImpl(long, bool, MemoryTracker*) @ 0xba4f9e7 in /usr/bin/clickhouse
3. MemoryTracker::allocImpl(long, bool, MemoryTracker*) @ 0xba4f471 in /usr/bin/clickhouse
4. DB::Memory<Allocator<false, false> >::alloc() @ 0xbaa2431 in /usr/bin/clickhouse
5. DB::WriteBufferFromFile::WriteBufferFromFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, unsigned long, int, unsigned int, char*, unsigned long) @ 0xbac178e in /usr/bin/clickhouse
6. DB::DiskLocal::writeFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, unsigned long, DB::WriteMode, DB::WriteSettings const&) @ 0x16312bf9 in /usr/bin/clickhouse
7. DB::DataPartStorageBuilderOnDisk::writeFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, unsigned long, DB::WriteSettings const&) @ 0x174b69c2 in /usr/bin/clickhouse
8. DB::DataPartsExchange::Fetcher::downloadBaseOrProjectionPartToDisk(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::IDataPartStorageBuilder>&, bool, DB::PooledReadWriteBufferFromHTTP&, DB::MergeTreeDataPartChecksums&, std::__1::shared_ptr<DB::Throttler>) const @ 0x174ceea7 in /usr/bin/clickhouse
9. DB::DataPartsExchange::Fetcher::downloadPartToDisk(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, bool, std::__1::shared_ptr<DB::IDisk>, DB::PooledReadWriteBufferFromHTTP&, unsigned long, DB::MergeTreeDataPartChecksums&, std::__1::shared_ptr<DB::Throttler>) @ 0x174cdd64 in /usr/bin/clickhouse
10. DB::DataPartsExchange::Fetcher::fetchPart(std::__1::shared_ptr<DB::StorageInMemoryMetadata const> const&, std::__1::shared_ptr<DB::Context const>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, DB::ConnectionTimeouts const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::Throttler>, bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::optional<DB::CurrentlySubmergingEmergingTagger>*, bool, std::__1::shared_ptr<DB::IDisk>) @ 0x174c5744 in /usr/bin/clickhouse
11. ? @ 0x17309a39 in /usr/bin/clickhouse
12. DB::StorageReplicatedMergeTree::fetchPart(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::StorageInMemoryMetadata const> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, bool, unsigned long, std::__1::shared_ptr<zkutil::ZooKeeper>, bool) @ 0x172587a8 in /usr/bin/clickhouse
13. DB::StorageReplicatedMergeTree::executeFetch(DB::ReplicatedMergeTreeLogEntry&, bool) @ 0x1724ace1 in /usr/bin/clickhouse
14. DB::ReplicatedMergeMutateTaskBase::executeImpl() @ 0x1773bd20 in /usr/bin/clickhouse
15. DB::ReplicatedMergeMutateTaskBase::executeStep() @ 0x1773af82 in /usr/bin/clickhouse
16. DB::MergeTreeBackgroundExecutor<DB::MergeMutateRuntimeQueue>::routine(std::__1::shared_ptr<DB::TaskRuntimeData>) @ 0xba0d7bb in /usr/bin/clickhouse
17. DB::MergeTreeBackgroundExecutor<DB::MergeMutateRuntimeQueue>::threadFunction() @ 0xba0d1f0 in /usr/bin/clickhouse
18. ThreadPoolImpl<ThreadFromGlobalPool>::worker(std::__1::__list_iterator<ThreadFromGlobalPool, void*>) @ 0xbb06e06 in /usr/bin/clickhouse
19. void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<ThreadFromGlobalPool::ThreadFromGlobalPool<void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda0'()>(void&&)::'lambda'(), void ()> >(std::__1::__function::__policy_storage const*) @ 0xbb08a15 in /usr/bin/clickhouse
```

**Additional context**
max_server_memory_usage 114894443315
![image](https://user-images.githubusercontent.com/103396/188755652-0d607d38-4751-48f5-987f-33967e729347.png)

**Comment by den-crane** [comment]
duplicate https://github.com/ClickHouse/ClickHouse/issues/39915
fixed 22.7.3.5 Backported in https://github.com/ClickHouse/ClickHouse/issues/40045: Fix big memory usage during fetches. Fixes https://github.com/ClickHouse/ClickHouse/issues/39915. https://github.com/ClickHouse/ClickHouse/pull/39990 ([Nikolai Kochetov](https://github.com/KochetovNicolai)).

https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelogs/v22.7.3.5-stable.md

**Comment by den-crane** [comment]
@fillest you need to update to 22.7.5.13
